# ФОРМИРОВАНИЕ ЦЕННОСТНОГО ПРЕДЛОЖЕНИЯ ДЛЯ КРУПНОГО СЕРВИСА ЧТЕНИЯ КНИГ ПО ПОДПИСКЕ

**Задача проекта:**

на основе данных, представленных в:
* **Таблица books**

Содержит данные о книгах:
* book_id — идентификатор книги
* author_id — идентификатор автора
* title — название книги
* num_pages — количество страниц
* publication_date — дата публикации книги
* publisher_id — идентификатор издателя


* **Таблица authors**

Содержит данные об авторах:
* author_id — идентификатор автора
* author — имя автора


* **Таблица publishers**

Содержит данные об издательствах:

* publisher_id — идентификатор издательства
* publisher — название издательства


* **Таблица ratings**

Содержит данные о пользовательских оценках книг:

* rating_id — идентификатор оценки
* book_id — идентификатор книги
* username — имя пользователя, оставившего оценку
* rating — оценка книги


* **Таблица reviews**

Содержит данные о пользовательских обзорах:

* review_id — идентификатор обзора
* book_id — идентификатор книги
* username — имя автора обзора
* text — текст обзора

**Проанализировать** базу данных,

**Сформулировать** выводы для формирования ценностного предложения.


## предобработка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                 db_config['pwd'],
                                                 db_config['host'],
                                                 db_config['port'],
                                                 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
# загружаем файл books

query_books = '''
            SELECT *
            FROM books
         '''
 
books = pd.io.sql.read_sql(query_books, con = engine)
display(books.head(3))
display(books.info())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

In [3]:
# загружаем файл authors

query_authors = '''
            SELECT *
            FROM authors
         '''
 
authors = pd.io.sql.read_sql(query_authors, con = engine)
display(authors.head(3))
display(authors.info())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

In [4]:
# загружаем файл publishers

query_publishers = '''
            SELECT *
            FROM publishers
         '''
 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
display(publishers.head(3))
display(publishers.info())

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

In [5]:
# загружаем файл ratings

query_ratings = '''
            SELECT *
            FROM ratings
         '''
 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
display(ratings.head(3))
display(ratings.info())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

In [6]:
# загружаем файл reviews

query_reviews = '''
            SELECT *
            FROM reviews
         '''
 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
display(reviews.head(3))
display(reviews.info())

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

### Выводы

* таблица **books** содержит 5 полей и 1 000 записей
* таблица **authors** содержит 2 поля и 636 записей
* таблица **publishers** содержит 2 поля и 340 записей
* таблица **ratings** содержит 4 поля и 6 456 записей
* таблица **reviews** содержит 4 поля и 2 793 записи

## Исследовальский анализ данных

### Расчет кол-ва книг, вышедших после 1 января 2000 года

In [7]:
# выполняем условие > '2000-01-01',
# считаем кол-во выпущенных книг в этот период

query_publication_year = '''
            SELECT COUNT (book_id) AS published_books
            FROM books
            WHERE publication_date > '2000-01-01';
         '''
 
publication_year = pd.io.sql.read_sql(query_publication_year, con = engine)
display(publication_year)


published_books
0              819

#### Выводы
* на период издания после 1 января 2000 года приходится 819 книг

### Расчет кол-ва обзоров и средней оценки для каждой книги

In [8]:
# объединяем таблицы book, reviews и ratings,
# аггругируем данные по  book_id,
# считаем кол-во отзывов и средний рейтинг

query_review_rating = '''
            SELECT b.book_id,
            title,
            COUNT(DISTINCT rev.review_id) AS review_number, AVG(r.rating) AS rating_avg
            FROM books AS b LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
            LEFT JOIN ratings AS r ON rev.book_id = r.book_id
            GROUP BY b.book_id
            ORDER BY review_number DESC;
            
         '''

review_rating = pd.io.sql.read_sql(query_review_rating, con = engine)
display(review_rating.head(10))
display(review_rating.tail(10))

book_id                                              title  review_number  \
0      948                            Twilight (Twilight  #1)              7   
1      963                                Water for Elephants              6   
2      734                                   The Glass Castle              6   
3      302  Harry Potter and the Prisoner of Azkaban (Harr...              6   
4      695  The Curious Incident of the Dog in the Night-Time              6   
5      696             The Da Vinci Code (Robert Langdon  #2)              6   
6      627                                      The Alchemist              6   
7      750                The Hobbit  or There and Back Again              6   
8      656                                     The Book Thief              6   
9      779  The Lightning Thief (Percy Jackson and the Oly...              6   

   rating_avg  
0    3.662500  
1    3.977273  
2    4.206897  
3    4.414634  
4    4.081081  
5    3.830508  
6    3.789474  
7    4.125000  
8    4.264151  
9    4.080645

book_id                                              title  \
990      413                                         Lysistrata   
991      599  Stone of Farewell (Memory  Sorrow  and Thorn  #2)   
992      535             Ranma ½  Vol. 1 (Ranma ½ (US 2nd)  #1)   
993      510                      Phantom (Sword of Truth  #10)   
994      387                               Leonardo's Notebooks   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     review_number  rating_avg  
990              1         4.0  
991              1         5.0  
992              1         4.0  
993              1         3.5  
994              0         NaN  
995              0         NaN  
996              0         NaN  
997              0         NaN  
998              0         NaN  
999              0         NaN

###### Выводы
* из 1000 книг, находящихся в базе данных, 6 книг оказались без информации об оценках и без обзоров
* ТОП 10 книг по по количеству отзывов - отзывы находятся в диапазоне о 6 до 7 шт на книгу
* Самы популярные по кол-ву отзывов книги имеют средний рейтинг в диапазоне 3,66 (Twilight (Twilight #1) - 7 отзывов) - 4,41 (Harry Potter and the Prisoner of Azkaban - 6 отзывов)

### Выявление издательства, которое выпустило наибольшее кол-во книг толще 50 стр.

In [9]:
# объединяем таблицы books и publishers,
# аггругируем данные по  publisher,
# считаем кол-во book_id с количеством страниц больше 50

query_publisher_top = '''
            SELECT p.publisher,
            COUNT(b.book_id) AS books_number
            FROM books AS b FULL JOIN publishers AS p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY p.publisher
            ORDER BY books_number DESC;
            
         '''

publisher_top = pd.io.sql.read_sql(query_publisher_top, con = engine)
display(publisher_top.head(10))

publisher  books_number
0             Penguin Books            42
1                   Vintage            31
2  Grand Central Publishing            25
3          Penguin Classics            24
4          Ballantine Books            19
5                    Bantam            19
6                   Berkley            17
7        St. Martin's Press            14
8             Berkley Books            14
9                     Delta            13

#### Выводы

* в ТОП 3 издательств по кол-ву выпущенных книг свыше 50 страниц (брошюры) в базе сервиса вошли:
1. Penguin Books = 42 книги
2. Vintage = 31 книга
3. Grand Central Publishing = 25 книг

### Выявление автора с самой высокой средней оценкой книг (книги с 50 и более оценками)

In [10]:
# Создать общую таблицу from books присоединив нужные нам таблицы - ratings (по book_id) и authors (по author_id)
# Группируем по book_id и author_id
# Добавляем фильтр HAVING COUNT(rating_id) >= 50 (достаточно одного раза)
# Cгруппировать по авторам и рассчитать среднюю оценку каждого (т.к. у некоторых несколько книг).

query_author_top = '''
            SELECT a.author,
            AVG(r.rating) AS rating_avg
            FROM books AS b FULL JOIN ratings AS r ON b.book_id = r.book_id
            FULL JOIN authors As a ON b.author_id = a.author_id
            WHERE b.book_id IN (
                                SELECT book_id
                                FROM ratings
                                GROUP BY book_id
                                HAVING COUNT(rating_id) >= 50
                                )
            GROUP BY a.author
            ORDER BY rating_avg DESC;
         '''

author_top = pd.io.sql.read_sql(query_author_top, con = engine)
display(author_top)

author  rating_avg
0                          J.K. Rowling/Mary GrandPré    4.287097
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.246914
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                          Lois Lowry    3.750000
10                                          Dan Brown    3.741259
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

#### Выводы
* 14 авторов попали в список авторов, чьи книги набрали более 50 оценок
* самую высокую среднюю оценку **4.29** книг среди них получили соавторы **J.K. Rowling/Mary GrandPré**

### Расчет среднего кол-ва обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
# объединяем таблицы books, ratings и author,
# делаем выборку книг с кол-вом оценок >= 50,
# аггрегируем данные по авторам, считаем средний рейтинг

query_review_avg = '''
            SELECT AVG(rev_count)
            FROM (
            SELECT username,
            COUNT(DISTINCT review_id) AS rev_count
            FROM reviews
            WHERE username IN (SELECT username
                                FROM ratings
                                GROUP BY username
                                HAVING COUNT(DISTINCT rating_id) > 50)
            GROUP BY username) AS final;
    
            
         '''

review_avg = pd.io.sql.read_sql(query_review_avg, con = engine)
display(review_avg)

avg
0  24.333333

#### Выводы
* в среднем пользователи, которые ставят больше 50 оценок, делают 24,33 обзора

## Рекомендации
* первичный **анализ базы** данных показал, что в ней **зарегистрировано**:
>
> **1 000** книг, 
>
> которые пренадлежат перу **636** авторов,
>
> были изданы **340** издателемя,
>
> получили **6 456** оценок и 
>
> **2 793** отзывов
>
* 82% книг приходятся изданными на период после 1 января 2000 года, что характеризует БД как достаточно современный контент. При этом 552 книги (55%) приходятся на период с 2003 по 2006 год. Это неплохая база для развития ценностного предложения с точки зрения выбора современной литературы. Но тогда рекомендуется нарастить селекцию и увеличить закупку произведений, изданных после 2006 года.
* Низкая доля (6 из 1 000) книг, оставшихся без оценок и отзывов, показывает достаточно высокий уровень лояльности клиентской базы. Рекомендуется использовать этот факт при формулировке ценностного предложения. 
* Лидерство издательства Penguin Books, которое заняло 1-е место с 44 книгами и 4-е с 24 в своей версии Penguin Books classics свидетельствует о преобладании семенйых ценностей в текущем ассортименте и у текущей аудитории сервиса. Тема вечная - рекомендуется к развитию.
* более половины авторов, получивших самые высокие средние оценки книг также издаются в Penguin Books, что подтверждает приверженность текущей аудитории сервиса к издательству и семейно - психологическим ценностям.
* Высокая активность пользователей (24 обзора среди тех, кто поставил более 50 оценок) подтверждает лояльность аудитории. 